orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -59,11 +62,56 @@ void FindLeafNodes(Graph& graph, std::vector<NodeArg*>& input_args) {
       }
     }
     if (!find_consumer_nodes && outputs.size() > 0) {
-      input_args.push_back(outputs[0]);
+      leaf_node_args.push_back(outputs[0]);
     }
   }
+
+  return leaf_node_args;
 };
 
+// This function converts tensor NodeArg to a boolean scalar so that last
+// backward RecordEvent doesn't block the early release of large gradient
+// tensors. If we connect gradient tensors directly to that RecordEvent,
+// we need a memory block (as large as a whole model) to store gradient
+// for each trainable tensor until the end of backward pass.
+//
+// The newly created bollean scalar may be appended to signal_args. If
+// signal_args is empty, the source of signal_args[i] would be tensor_args[i].
+void ConvertTensorToBoolSignal(
+  Graph& graph,
+  const std::vector<NodeArg*>& tensor_args,
+  std::vector<NodeArg*>& signal_args) {
+
+  for (auto tensor_arg: tensor_args) {
+    // Declare the scalar signal this "tensor_arg" will be converted into.
+    auto signal_arg = &CreateTypedNodeArg(
+      graph,
+      ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+      "signal_" + tensor_arg->Name()
+    );
+
+    // Add the new scalar to user-specified vector.
+    signal_args.push_back(signal_arg);
+
+    // Add tensor-to-scalar conversion node.
+    const auto name = graph.GenerateNodeName("tensor_to_scalar_signal");
+    std::vector<NodeArg*> input_args{tensor_arg};
+    std::vector<NodeArg*> output_args{signal_arg};
+    graph.AddNode(
+      name,
+      "Group",
+      "",
+      input_args,
+      output_args,
+      nullptr,
+      kMSDomain);
+
+    // TODO: remove this break after changining topological sort to depth-first
+    // from inputs.
+    break;
```

Comment:
>break; [](start = 4, length = 6)

If we move the graph.AddNode() for Group node outside the for loop, maybe we don't need this break?

Reply:
No. This is a temporal solution. We will remove this break in the future.

---
In reply to: [439679419](https://github.com/microsoft/onnxruntime/pull/4204#discussion_r439679419) [](ancestors = 439679419)

Reply:
The break here will make just the first item in tensor_args connected with a Group node (And the rest keep dangling) . Could you double check if this is expected?

---
In reply to: [439679419](https://github.com/microsoft/onnxruntime/pull/4204#discussion_r439679419) [](ancestors = 439679419)